device/pools.go

```diff
@@ -57,11 +65,38 @@ func (device *Device) PopulatePools() {
 	})
 }
 
+func (device *Device) GetOutboundElementsSlice() *[]*QueueOutboundElement {
+	return device.pool.outboundElementsSlice.Get().(*[]*QueueOutboundElement)
+}
+
+func (device *Device) PutOutboundElementsSlice(s *[]*QueueOutboundElement) {
+	for i := range *s {
+		(*s)[i] = nil
+	}
+	*s = (*s)[:0]
+	device.pool.outboundElementsSlice.Put(s)
+}
+
+func (device *Device) GetInboundElementsSlice() *[]*QueueInboundElement {
+	return device.pool.inboundElementsSlice.Get().(*[]*QueueInboundElement)
+}
+
+func (device *Device) PutInboundElementsSlice(s *[]*QueueInboundElement) {
+	for i := range *s {
+		(*s)[i] = nil
+	}
+	*s = (*s)[:0]
+	device.pool.inboundElementsSlice.Put(s)
+}
+
 func (device *Device) GetMessageBuffer() *[MaxMessageSize]byte {
 	return device.pool.messageBuffers.Get().(*[MaxMessageSize]byte)
 }
 
 func (device *Device) PutMessageBuffer(msg *[MaxMessageSize]byte) {
+	if msg == nil {
+		panic("msg is nil")
```

Comment:
Is this something you guys hit or is it just best practice when dealing with pools?

Reply:
We hit this is in a very early phase of development before guarding `defer` cleanups that recycled nil pointers. Not intended to be left here, I've removed it in https://github.com/WireGuard/wireguard-go/pull/64/commits/61b7b39ea3ab4c9fb9b20dcd4038f82db4e898ed.